### PR TITLE
Implement 1cycle learning rate policy

### PR DESCRIPTION
**Pull request by mjacar**
# What is this?
This is an implementation of the 1cycle learning rate policy as implemented in [the fastai library](https://github.com/fastai/fastai/blob/79ecfd5c60087766c6f1952e674f98470d488069/fastai/callbacks/one_cycle.py) and as initially introduced in the paper [Super-Convergence: Very Fast Training of Neural Networks Using Large Learning Rates](https://arxiv.org/abs/1708.07120).

This was requested in https://github.com/pytorch/pytorch/issues/19129. It has been suggested in both https://github.com/pytorch/pytorch/issues/19129 and https://github.com/pytorch/pytorch/pull/2016 that the 1cycle policy could be implemented in terms of the existing `CyclicLR` scheduler. However, that isn't actually true. In particular, the 1cycle policy calls for annealing from the maximum learning rate down to a minimum learning rate that is multiple orders of magnitude lower than the initial learning rate. In the `CyclicLR` scheduler, the initial learning rate and the minimum learning rate are the same.
# Why is this important?
It has been shown that the 1cycle learning rate policy can speed up convergence tremendously. In one example from [the original paper](https://arxiv.org/pdf/1708.07120.pdf), a model trained on MNIST could reach 99.03% accuracy after 85 epochs using previous techniques while the same model could reach 99.3% accuracy in just 12 epochs using the 1 cycle policy.
# How was this tested?
Test cases were added to `test_optim.py`.

### CODE REVIEW

**Comment by mjacar** [comment]
The linter is still complaining about the `\:` sequence on line 864 of `lr_scheduler.py`. However, unless my eyes are deceiving me, that appears to be exactly the same as the `\:` sequence on line 707. Am I not seeing something or did line 707 get grandfathered in somehow?

**Comment by mjacar** [comment]
@soumith Fair enough if it's an inherently unanswerable question, but is there any rough ETA on when this might get triaged?

**Comment by soumith** [comment]
cc: @vincentqb can you review this and get it to completion if it makes sense. Use the guidelines that I shared with you separately.

@mjacar again, sorry for the delay -- we have Vincent now, who is owning the reviews for optimizers and LR schedulers (LR schedulers are sufficently owned by @ezyang as well :) )

**Comment by Ir1d** [comment]
@vincentqb Hi, any update on this PR? Really looking forward to this 1cycle policy. Thx in advance.

**Comment by vincentqb** [comment]
Before moving forward, I'd like to make sure it is well-established (as mentioned in the [guidelines](https://github.com/pytorch/pytorch/issues/21988#issuecomment-503822859)). Do you have other applications/demonstrations of 1cycle beyond the one mentioned in the original paper?

**Comment by mjacar** [comment]
@vincentqb 

The popularity of this method comes largely from Jeremy Howard's [fast.ai MOOC course](https://course.fast.ai/) where he has taught this method as essentially the go-to method for training models quickly for about a year now. As such, there are a number of demonstrations of the method's usefulness across a variety of different domains and model architectures in just the course notes themselves.

Just going through the course notes and finding all references to `fit_one_cycle` yields the following:
[lesson1-pets](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson1-pets.ipynb): CNN image classifier on the [Oxford-IIIT Pet dataset](https://www.robots.ox.ac.uk/~vgg/data/pets/)
[lesson2-download](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson2-download.ipynb): Another CNN image classifier on a custom dataset
[lesson3-planet](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-planet.ipynb): Multi-label CNN image classifier on the [Planet dataset](https://www.kaggle.com/c/planet-understanding-the-amazon-from-space)
[lesson3-imdb](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-imdb.ipynb): Finetuned an RNN language model and then used that as a basis for transfer learning for training an RNN text classifier on the [Large Movie Review Dataset](http://ai.stanford.edu/~amaas/data/sentiment/)
[lesson3-head-pose](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-head-pose.ipynb): CNN head pose estimator on the [Biwi Kinect Head Pose dataset](https://data.vision.ee.ethz.ch/cvl/gfanelli/head_pose/head_forest.html#db)
[lesson3-camvid](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-camvid.ipynb): U-Net image segmentation model on the [Camvid dataset](http://mi.eng.cam.ac.uk/research/projects/VideoRec/CamVid/)
[lesson3-camvid-tiramisu](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-camvid-tiramisu.ipynb): Another U-Net based image segmentation model on the Camvid dataset
[lesson4-collab](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson4-collab.ipynb): Collaborative filtering model on the [Movielens 100k dataset](https://grouplens.org/datasets/movielens/100k/)
[lesson6-rossmann](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson6-rossmann.ipynb): Feedforward model for the tabular [Rossman dataset](https://www.kaggle.com/c/rossmann-store-sales)
[lesson6-pets-more](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson6-pets-more.ipynb): Another CNN pet classifier on the Oxford Pet dataset
[lesson7-superres](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson7-superres.ipynb): U-Net based super-resolution model
[lesson7-superres-imagenet](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson7-superres-imagenet.ipynb): Another U-Net based super-resolution model
[lesson7-superres-gan](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson7-superres-gan.ipynb): Another U-Net based super-resolution model and another CNN based image classifier
[lesson7-resnet-mnist](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson7-resnet-mnist.ipynb): More CNN image classifiers
[lesson-7-human-numbers](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson7-human-numbers.ipynb): Some feedforward and RNN models

Plus, I would argue the fact that this has come up already before in https://github.com/pytorch/pytorch/issues/19129 and https://github.com/pytorch/pytorch/pull/2016 gives more credence to the argument of this method's notability/usefulness.

**Comment by mjacar** [comment]
@vincentqb Any ETA on when this might get looked at? Is resolving https://github.com/pytorch/pytorch/pull/23306 a necessary prerequisite?

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -759,3 +759,220 @@ def step(self, epoch=None):
...
+        elif total_steps is not None:
+            if total_steps <= 0 or not isinstance(total_steps, int):
+                raise ValueError("Expected non-negative integer total_steps, but got {}".format(total_steps))
+            self.total_steps = total_steps
+        else:
+            if num_epochs <= 0 or not isinstance(num_epochs, int):
+                raise ValueError("Expected non-negative integer num_epochs, but got {}".format(num_epochs))
+            if not hasattr(train_dl, '__len__'):
+                raise ValueError("len is not defined over given train_dl")
+            self.total_steps = num_epochs * len(train_dl)
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py]: Is this the only place where data loader is used? If so, would it be better to pass as a parameter the length of the data instead of the loader itself?

**Review by vincentqb — COMMENTED** [review]
Thanks for the information above. Yes, I recommend that we proceed with this PR. :)

* As you pointed out, #23306 is planning to introduce chainable schedulers. As far as I can see, this is not a chainable scheduler, and overrides the current learning rate value, is that correct? 
* It doesn't seem like the `epoch` parameter for `step` needed, is that correct? We are moving away from it as part of #23306.

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (mjacar) [reply torch/optim/lr_scheduler.py]: Yeah, the data loader is only used to help compute the number of total steps in the cycle. Passing only the length of the data loader instead of the data loader itself is an entirely valid approach. I went with this approach because I thought it might arguably be slightly more straightforward for the programmer but I'm certainly not married to the idea whatsoever.

**Comment by mjacar** [comment]
```As you pointed out, #23306 is planning to introduce chainable schedulers. As far as I can see, this is not a chainable scheduler, and overrides the current learning rate value, is that correct?```

Yes. In fact, I'm not really sure what making a 1cycle policy "chainable" would even mean necessarily. It's a specific policy and not really a composable abstraction in any sense that is obvious to me. Do all schedulers need to be chainable?

 ```It doesn't seem like the epoch parameter for step needed, is that correct? We are moving away from it as part of #23306.```

I believe that's correct.

**Comment by vincentqb** [comment]
> Yes. In fact, I'm not really sure what making a 1cycle policy "chainable" would even mean necessarily. It's a specific policy and not really a composable abstraction in any sense that is obvious to me. Do all schedulers need to be chainable?

No, they don't need to all be chainable. I just wanted to make sure we were implementing such a form if available. We should mention this is not chainable in the docstring though.

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Good, let's go with length then. It is more explicit, and I could imagine cases in which length has to be inferred (and so leaving the user in control would be more appropriate).

Comment (mjacar) [reply torch/optim/lr_scheduler.py]: Sounds good. Is the plan for you to more fully review/commandeer this PR once the chainable work is complete?

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -759,3 +759,220 @@ def step(self, epoch=None):
...
+            return param
+        else:
+            return [param] * len(optimizer.param_groups)
+
+    def _annealing_cos(self, start, end, pct):
+        "Cosine anneal from `start` to `end` as pct goes from 0.0 to 1.0."
+        cos_out = math.cos(math.pi * pct) + 1
+        return end + (start - end) / 2.0 * cos_out
+
+    def _annealing_linear(self, start, end, pct):
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py]: How diverse is the set of functions we should expect it? I like the fact that a user can easily just add more functions as desired in this implementation. Could we end up combining schedulers in some way?

File: test/test_optim.py [hunk 3]
```diff
@@ -1013,6 +1013,58 @@ def test_cycle_lr_cycle_momentum_fail_with_momentumless_optimizer(self):
...
+    def test_onecycle_lr_invalid_pct_start(self):
+        with self.assertRaises(ValueError):
+            scheduler = OneCycleLR(self.opt, max_lr=1e-3, total_steps=10, pct_start=1.1)
+
+    def test_onecycle_lr_cannot_calculate_total_steps(self):
+        with self.assertRaises(ValueError):
+            scheduler = OneCycleLR(self.opt, max_lr=1e-3)
+
+    def test_onecycle_lr_linear_annealing(self):
+        lr_target = [1, 13, 25, 21.5, 18, 14.5, 11, 7.5, 4, 0.5]
```
Comment (vincentqb) [inline test/test_optim.py]: When the paper says "several orders of magnitude lower", I expected more than 3 :) but it's fine for testing

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Let's not worry about the other PR (#24352) for now: I can take care of that part :) Besides the comments I left, the PR looks good to me!

File: test/test_optim.py (hunk 3 above)
Comment (mjacar) [reply test/test_optim.py]: Yeah, I just cherry-picked the numbers so that the math was easy to do in my head so I could actually validate correctness.

File: torch/optim/lr_scheduler.py (hunk 2 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: Great.

File: test/test_optim.py (hunk 3 above)
Comment (vincentqb) [reply test/test_optim.py]: Sounds good to me :)

**Comment by mjacar** [comment]
@vincentqb I believe I've responded to all unresolved feedback with the latest changes. Feel free to take a look whenever you get the chance.

File: torch/optim/lr_scheduler.py [hunk 4]
```diff
@@ -759,3 +759,223 @@ def step(self, epoch=None):
...
+        last_epoch (int): The index of the last batch. This parameter is used when
+            resuming a training job. Since `step()` should be invoked after each
+            batch instead of after each epoch, this number represents the total
+            number of *batches* computed, not the total number of epochs computed.
+            When last_epoch=-1, the schedule is started from the beginning.
+            Default: -1
+
+    Example:
+        >>> data_loader = torch.utils.data.DataLoader(...)
+        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
```
Comment (aslamplr) [inline torch/optim/lr_scheduler.py]: Suggestion to make the example for meaningful. Since the `min_lr` is calculated from `initial_lr` which is the optimizer's `lr` and should be a lower value than the max_lr of `OneCycleLR`. This would be especially helpful for people who read the examples and try to understand the working. If possible please add a note about this as well. 
```suggestion
        >>> optimizer = torch.optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
```

Thank you for the PR

**Comment by vincentqb** [comment]
Landing from #25325
